src/aspire/io/starfile.py

```diff
@@ -183,22 +185,46 @@ def save_star(image_source, starfile_filepath, batch_size=1024, overwrite=False)
     df['_rlnImageName'] = ''
 
     with open(starfile_filepath, 'w') as f:
-        for i_start in np.arange(0, image_source.n, batch_size):
-
-            i_end = min(image_source.n, i_start + batch_size)
-            num = i_end - i_start
-
-            mrcs_filename = os.path.splitext(os.path.basename(starfile_filepath))[0] + f'_{i_start}_{i_end}.mrcs'
+        if save_mode == 'single':
+            # save all images into one single mrc file
+            mrcs_filename = os.path.splitext(os.path.basename(starfile_filepath)
+                                            )[0] + f'_{0}_{image_source.n-1}.mrcs'
             mrcs_filepath = os.path.join(
                 os.path.dirname(starfile_filepath),
                 mrcs_filename
             )
+            df['_rlnImageName'][0: image_source.n] = pd.Series(['{0:06}@{1}'.format(j + 1, mrcs_filepath)
+                                                                for j in range(image_source.n)])
+            with mrcfile.new_mmap(mrcs_filepath, shape=(image_source.n, image_source.L,
+                                                        image_source.L), mrc_mode=2, overwrite=overwrite) as mrc:
+                for i_start in np.arange(0, image_source.n, batch_size):
+                    i_end = min(image_source.n, i_start + batch_size)
+                    num = i_end - i_start
+                    logger.info(f'Saving ImageSource[{i_start}-{i_end-1}] to {mrcs_filepath}')
+                    mrc.data[i_start:i_end, :, :] = np.swapaxes(image_source.images(
+                        start=i_start, num=num).data.astype('float32'), 0, 2)
+            mrc.close()
 
-            logger.info(f'Saving ImageSource[{i_start}-{i_end}] to {mrcs_filepath}')
-            im = image_source.images(start=i_start, num=num)
-            im.save(mrcs_filepath, overwrite=overwrite)
-
-            df['_rlnImageName'][i_start: i_end] = pd.Series(['{0:06}@{1}'.format(j + 1, mrcs_filepath) for j in range(num)])
-
+        else:
+            # save all images into multiple mrc files in batch size
+            for i_start in np.arange(0, image_source.n, batch_size):
+
+                i_end = min(image_source.n, i_start + batch_size)
+                num = i_end - i_start
+                mrcs_filename = os.path.splitext(os.path.basename(starfile_filepath)
```

Comment:
Don't split the line just before parenthesis.

Reply:
Done.